tests/time_integration_with_constraints.cpp

```diff
@@ -0,0 +1,366 @@
+    mfem::Vector duf_dt = du_dt;
+    duf_dt[1] = uf[1] = 0;
+
+    mfem::Vector f = fext(t) - M * duc_dt(t) - K * uc(t) - K * uf;
+    f[1]           = 0;
+
+    du_dt = invT * f;
+
+    dt_prev = dt;
+  };
+
+  ode.SetMethod< mfem::ImplicitMidpointSolver >();
+
+//  auto time_integrator = mfem::ImplicitMidpointSolver();
+//  time_integrator.Init(ode);
+
+  mfem::Vector u = uc(0);
+  double       t = 0.0;
+  for (int i = 0; i < num_steps; i++) {
+    ode.Step(u, t, dt);
+    u[1] = uc(t)[1];
+  }
+
+  mfem::Vector solution(3);
+  solution[0] = -0.1443913076373983;
+  solution[1] = 0.0;
+  solution[2] = 0.04968869236260167;
+
+  mfem::Vector error = u - solution;
+
+  std::cout << error << std::endl;
+
+  return error.Norml2();
+}
+
+double first_order_ode_test(int num_steps)
+{
+  // clang-format off
+  mfem::DenseMatrix M(3, 3);
+  M(0, 0) = 2; M(0, 1) = 1; M(0, 2) = 0;
+  M(1, 0) = 1; M(1, 1) = 4; M(1, 2) = 1;
+  M(2, 0) = 0; M(2, 1) = 1; M(2, 2) = 2;
+
+  double k = 100;
+  mfem::DenseMatrix K(3, 3);
+  K(0, 0) =  k; K(0, 1) =    -k; K(0, 2) =  0;
+  K(1, 0) = -k; K(1, 1) = 2 * k; K(1, 2) = -k;
+  K(2, 0) =  0; K(2, 1) =    -k; K(2, 2) =  k;
+  // clang-format on
+
+  double gamma = 0.5;
+  double tmax  = 1.0;
+  double dt    = tmax / num_steps;
+
+  mfem::DenseMatrix invH = M;
+  mfem::DenseMatrix invT = M;
+  invT.AddMatrix(dt * gamma, K, 0, 0);
+
+  for (int i = 0; i < 3; i++) {
+    invH(i, 1) = invH(1, i) = (i == 1);
+    invT(i, 1) = invT(1, i) = (i == 1);
+  }
+
+  invH.Invert();
+  invT.Invert();
+
+  double       t      = 0.0;
+  mfem::Vector u      = uc(t);
+  mfem::Vector du_dt  = invH * (fext(t) - K * u - M * duc_dt(t));
+  du_dt[1]            = duc_dt(t)[1];
+  mfem::Vector uf     = u;
+  mfem::Vector duf_dt = du_dt;
+  uf[1] = duf_dt[1] = 0.0;
+
+  for (int i = 0; i < num_steps; i++) {
+    t += dt;
+
+    mfem::Vector f = fext(t) - M * duc_dt(t) - K * uc(t) - K * (uf + (1.0 - gamma) * duf_dt * dt);
+    f[1]           = 0;
+
+    uf += (1.0 - gamma) * duf_dt * dt;
+    duf_dt = invT * f;
+    uf += gamma * duf_dt * dt;
+
+    u = uf + uc(t);
+  }
+
+  mfem::Vector solution(3);
+  solution[0] = -0.1443913076373983;
+  solution[1] = 0.0;
+  solution[2] = 0.04968869236260167;
+
+  mfem::Vector error = u - solution;
+
+  return error.Norml2();
+}
+
+double second_order_ode_test(int num_steps)
+{
+  // clang-format off
+  mfem::DenseMatrix M(3, 3);
+  M(0, 0) = 2; M(0, 1) = 1; M(0, 2) = 0;
+  M(1, 0) = 1; M(1, 1) = 4; M(1, 2) = 1;
+  M(2, 0) = 0; M(2, 1) = 1; M(2, 2) = 2;
+
+  mfem::DenseMatrix C(3, 3);
+  C(0, 0) =  1; C(0, 1) = -1; C(0, 2) =  0;
+  C(1, 0) = -1; C(1, 1) =  2; C(1, 2) = -1;
+  C(2, 0) =  0; C(2, 1) = -1; C(2, 2) =  1;
+
+  double k = 100;
+  mfem::DenseMatrix K(3, 3);
+  K(0, 0) =  k; K(0, 1) =    -k; K(0, 2) =  0;
+  K(1, 0) = -k; K(1, 1) = 2 * k; K(1, 2) = -k;
+  K(2, 0) =  0; K(2, 1) =    -k; K(2, 2) =  k;
+  // clang-format on
+
+  double beta  = 1.0 / 6.0;
+  double gamma = 1.0 / 2.0;
```

Comment:
`gamma` gets highlighted differently for me because `gamma` is a deprecated C function - I don't think we need to worry about a collision here, though.